### PR TITLE
feat: added Data Residency support

### PR DESCRIPTION
**Pull request by desusai7**
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
## Checklist:
- [x] Version updated in `README`
- [x] Version updated in `RSConstants.m`(v1)/`RSConstant.swift`(v2)
- [x] Version updated in `Rudder.xcodeproj`
- [x] Version updated in `Rudder.podspec`
- [x] `CHANGELOG` Updated
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation
- [ ] Passed `pod lib lint --no-clean --allow-warnings`

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-ios/203)

### CODE REVIEW

**Review by pallabmaiti — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSConfig.m [hunk 1]
```diff
@@ -34,11 +35,12 @@ - (instancetype)init
 }
 
 - (instancetype)init:(NSString *) dataPlaneUrl
+withDataResidencyServer: (RSDataResidencyServer) dataResidencyServer
```
Comment (pallabmaiti) [inline Sources/Classes/RSConfig.m:38]:
```suggestion
dataResidencyServer: (RSDataResidencyServer) dataResidencyServer
```

File: Sources/Classes/RSServerConfigManager.m [hunk 2]
```diff
@@ -114,6 +114,15 @@ - (RSServerConfigSource *)_parseConfig:(NSString *)configStr {
             destination.destinationConfig = [destinationDict objectForKey:@"config"];
             [destinations addObject:destination];
         }
+        
+        NSDictionary *dataResidencyUrls = [sourceDict objectForKey:@"dataPlaneUrls"];
+        for (NSString *key in [dataResidencyUrls allKeys]) {
+            if([[key lowercaseString] isEqualToString:@"us"]) {
```
Comment (pallabmaiti) [inline Sources/Classes/RSServerConfigManager.m:120]: This parsing doesn't make sense here. Again you are finding the URL in RSServerConfigSource. Simply assign the `dataResidencyUrls` to `source.dataResidencyUrls`. Parse the urls in RSServerConfigSource and use `switch`.

**Review by pallabmaiti — CHANGES_REQUESTED** [review]

File: Sources/Classes/RSServerConfigManager.m (hunk 2 above)
Comment (desusai7) [reply Sources/Classes/RSServerConfigManager.m:120]: The reason why I am parsing here is that irrespective of whether we received the region keys from `sourceConfig` in `Capitals`  (or) `LowerCase` we are retrieving them and then in the case of RSServerConfigSource we are simply Storing the `Enum` to `URL` mapping. 

We are not iterating through the entire dictionary again in the RSServerConfigSource, we are just checking if a `RSDataResidencyServer`  enum exists, if yes returning the data plane URL.

We are cleaning/processing the data from the sourceConfig in the `_parseConfig` of `RSServerConfigManager` and then making it ready to use it for the `RSServerConfigSource`.

Please let me know, if this makes sense now

File: Sources/Classes/RSConfig.m (hunk 1 above)
Comment (desusai7) [reply Sources/Classes/RSConfig.m]: Made changes accordingly

File: Sources/Classes/RSServerConfigManager.m (hunk 2 above)
Comment (pallabmaiti) [reply Sources/Classes/RSServerConfigManager.m:120]: > The reason why I am parsing here is that irrespective of whether we received the region keys from sourceConfig in Capitals (or) LowerCase we are retrieving them and then in the case of RSServerConfigSource we are simply Storing the Enum to URL mapping.
Here, you are parsing & storing again; it doesn't make sense.

Comment (desusai7) [reply Sources/Classes/RSServerConfigManager.m]: As discussed,  handling only the lowercase region values from the sourceConfig for now, and hence remove the parsing logic completely.

**Review by pallabmaiti — APPROVED** [review]

**Review by pallabmaiti — APPROVED** [review]

File: Sources/Classes/RSServerConfigSource.m [hunk 3]
```diff
@@ -23,4 +24,26 @@ - (void)addDestination:(RSServerDestination *)destination {
     [self.destinations addObject:destination];
 }
 
+- (NSString *) getDataResidencyUrl:(RSDataResidencyServer) residency {
+    NSArray * residenceDataPlanes;
+    switch(residency) {
+        case EU:
+            residenceDataPlanes = [self.dataPlanes objectForKey:@"EU"];
```
Comment (pallabmaiti) [inline Sources/Classes/RSServerConfigSource.m:31]: Why break is not here? I don't understand the logic here. Can we have a call?

File: Sources/Classes/RSServerConfigSource.m [hunk 4]
```diff
@@ -23,4 +24,26 @@ - (void)addDestination:(RSServerDestination *)destination {
...
+            if (residenceDataPlanes != nil)
+                break;
+            residenceDataPlanes = [self.dataPlanes objectForKey:@"US"];
+    }
+    
+    if(residenceDataPlanes == nil)
+        return nil;
+    for (NSDictionary* residenceDataPlane in residenceDataPlanes) {
+        if([[residenceDataPlane objectForKey:@"default"] boolValue]) {
+            NSLog(@"Data type is %@",[[residenceDataPlane objectForKey:@"default"] class]);
```
Comment (pallabmaiti) [inline Sources/Classes/RSServerConfigSource.m:42]: Please remove the NSLog.

**Review by pallabmaiti — CHANGES_REQUESTED** [review]
I can see 3 bugs for this PR on [Sonarqube](https://sonarcloud.io/summary/new_code?id=rudderlabs_rudder-sdk-ios&pullRequest=203). Can you resolve those?

File: Sources/Classes/RSServerConfigSource.m (hunk 4 above)
Comment (desusai7) [reply Sources/Classes/RSServerConfigSource.m:42]: Done.

File: Sources/Classes/RSServerConfigSource.m (hunk 3 above)
Comment (desusai7) [reply Sources/Classes/RSServerConfigSource.m:31]: Added a comment to explain the logic here

**Review by pallabmaiti — APPROVED** [review]

**Review by pallabmaiti — DISMISSED** [review]

**Comment by pallabmaiti** [comment]
@desusai7 I can see 1 bug in [Sonarcloud](https://sonarcloud.io/summary/new_code?id=rudderlabs_rudder-sdk-ios&pullRequest=203). Can you take care that?

**Comment by desusai7** [comment]
@pallabmaiti fixed it now

**Review by pallabmaiti — APPROVED** [review]

**Pull request merged**